Add the PgSQL driver

derrabus opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | N/A

#### Summary

This PR introduces a new Postgres driver that leverages PHP's `pgsql` extension. Postgres is currently the last database that we only support through PDO. This driver closes the gap.

`ext-pgsql` has a rather verbose API. Unlike other database extensions, it does not even bother to cast result types: It returns everything as a string. Except for `null` values: Those are really `null`!

However, we get information on the database type of all result fields, so we are able to perform those casts ourselves. This is why I decided to perform very basic type mappings in the driver already. Otherwise, the result sets produced by `ext-pgsql` would have been a permanent edge case: Booleans come as `'t'` and `'f'`, binary data is returned as hex string prefixed with `0x`. Not casting those would cause a lot of functional tests to fail.

The type casts that I perform in the driver should be similar to what `PDO_pgsql` does internally as well. That should make a potential switch from PDO to the native extension rather seamless.

In PHP 8.1, all resource types used by the extensions have been changed to objects. This is why you will see a couple of type annotations with a union of `resource` and some class from the `PgSql` namespace. This PR currently targets DBAL 3.6 and PHP 7.4. Once this code hits the 4.x branch, we can simplify a lot of type declarations.

The one big limitation though is that I haven't implemented support for loading BLOBs from resources yet. That topic is a bit complex, so I'd like to work on that in a follow-up. For now, I think we can ship this driver without BLOB support.

derrabus committed:
Add the PgSQL driver

derrabus commented on psalm.xml.dist:
```diff
@@ -19,7 +19,6 @@
         <file name="vendor/jetbrains/phpstorm-stubs/ibm_db2/ibm_db2.php" />
         <file name="vendor/jetbrains/phpstorm-stubs/mysqli/mysqli.php" />
         <file name="vendor/jetbrains/phpstorm-stubs/oci8/oci8.php" />
-        <file name="vendor/jetbrains/phpstorm-stubs/pgsql/pgsql.php" />
```
Psalm actually runs better using its own `ext-pgsql` stubs. 🧐

greg0ire commented on src/Driver/PgSQL/Connection.php:
```diff
@@ -0,0 +1,130 @@
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource of an instance of %s, got %s.',
```
```suggestion
                'Expected connection to be a resource or an instance of %s, got %s.',
```

greg0ire commented on src/Driver/PgSQL/Connection.php:
```diff
@@ -0,0 +1,130 @@
+        return $this->query('SELECT LASTVAL()')->fetchOne();
+    }
+
+    public function beginTransaction(): bool
```
Should these be documented with `@return true`?

derrabus replied:
We could do this for most of the non-PDO drivers if we want. On 4.0, we've switched all of them to void.

greg0ire commented on src/Driver/PgSQL/Result.php:
```diff
@@ -0,0 +1,246 @@
+    {
+        if (! is_resource($result) && ! $result instanceof PgSqlResult) {
+            throw new TypeError(sprintf(
+                'Expected result to be a resource of an instance of %s, got %s.',
```
```suggestion
                'Expected result to be a resource or an instance of %s, got %s.',
```

greg0ire commented on src/Driver/PgSQL/Statement.php:
```diff
@@ -0,0 +1,157 @@
+    {
+        if (! is_resource($connection) && ! $connection instanceof PgSqlConnection) {
+            throw new TypeError(sprintf(
+                'Expected connection to be a resource of an instance of %s, got %s.',
```
```suggestion
                'Expected connection to be a resource or an instance of %s, got %s.',
```

derrabus commented:
Running the ORM test suite uncovered a nice little bug in the type mapping logic. 😓 On it.

derrabus commented:
I've added `Doctrine\DBAL\Tests\Functional\Driver\PgSQL\ResultTest::testTypeConversionWithDuplicateFieldNames()` to reproduce the problem found by the ORM's test suite and fixed the issue.

The ORM's test suite passes now, but it's significantly slower with the new driver than it is with PDO. 😕 

But I can investigate this in a follow-up, I guess.

greg0ire commented on src/Driver/PgSQL/Driver.php:
```diff
@@ -0,0 +1,95 @@
+
+final class Driver extends AbstractPostgreSQLDriver
+{
+    /** @inheritdoc */
```
I don't recall the reason but this is usually done like this:

```suggestion
    /** {@inheritDoc} */
```

There are more occurrences in this PR.

derrabus replied:
Looks like I did this wrong on the SQLite3 driver as well.

greg0ire commented on src/Driver/PgSQL/Driver.php:
```diff
@@ -0,0 +1,95 @@
+        );
+
+        return implode(' ', array_map(
+            static fn ($value, string $key) => sprintf('%s=\'%s\'', $key, addslashes($value)),
```
```suggestion
            static fn ($value, string $key) => sprintf("%s='%s'", $key, addslashes($value)),
```

greg0ire commented on src/Driver/PgSQL/Exception/UnknownParameter.php:
```diff
@@ -0,0 +1,18 @@
+    public static function new(string $param): self
+    {
+        return new self(
+            sprintf('Could not find parameter %d in the SQL statement', $param),
```
Should it be `%s` here?

```suggestion
            sprintf('Could not find parameter %s in the SQL statement', $param),
```

derrabus replied:
Yes

greg0ire commented:
> but it's significantly slower with the new driver than it is with PDO

Should we somehow document the new driver as experimental because of this?

derrabus commented:
Done.

andrew-demb commented:
> The ORM's test suite passes now, but it's significantly slower with the new driver than it is with PDO. 😕

It may be related to "real" prepared statements, in comparison to emulated prepared statements in PDO.

derrabus commented:
You mean because those real prepared statements mean an extra round-trip to the Posgres server? But that is nevertheless desirable, isn't it? I'm not sure I want to resolve prepared statements in the driver. 🙈
